geom/twkb_write.go

```diff
@@ -0,0 +1,535 @@
+package geom
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+)
+
+// MarshalTWKB accepts a geometry and generates the corresponding TWKB byte slice.
+func MarshalTWKB(geom Geometry,
+	hasZ, hasM bool,
+	precXY, precZ, precM int,
+	hasSize, hasBBox, closeRings bool,
+	idList []int64,
+) ([]byte, error) {
+	w := newTWKBWriter(hasZ, hasM, precXY, precZ, precM, hasSize, hasBBox, closeRings, idList)
+	if err := w.writeGeometry(geom); err != nil {
+		return nil, err
+	}
+	return w.twkb, nil
+}
+
+// TWKBWriter holds all state information needed for generating TWKB data
+// including information such as the last reference point used in coord deltas.
+type TWKBWriter struct {
+	twkb []byte
+
+	kind  twkbGeometryType
+	ctype CoordinatesType
+
+	dimensions int
+	precXY     int
+	hasZ       bool
+	hasM       bool
+	precZ      int
+	precM      int
+	scalings   [twkbMaxDimensions]float64
+
+	hasBBox bool
+	hasSize bool
+	hasIDs  bool
+	hasExt  bool
+	isEmpty bool
+
+	refpoint [twkbMaxDimensions]int64
+
+	bboxValid bool
+	bboxMin   [twkbMaxDimensions]int64
+	bboxMax   [twkbMaxDimensions]int64
+
+	idList []int64
+
+	closeRings bool
+}
+
+func newTWKBWriter(
+	hasZ, hasM bool,
+	precXY, precZ, precM int,
+	hasSize, hasBBox, closeRings bool,
+	idList []int64,
+) *TWKBWriter {
+	w := TWKBWriter{
+		hasSize:    hasSize,
+		hasBBox:    hasBBox,
+		closeRings: closeRings,
+	}
+
+	w.precXY = precXY
+	w.scalings[0] = math.Pow10(precXY)
+	w.scalings[1] = w.scalings[0]
+	w.dimensions = 2
+
+	if hasZ {
+		w.hasZ = true
+		w.precZ = precZ
+		w.scalings[w.dimensions] = math.Pow10(precZ)
+		w.dimensions++
+	}
+
+	if hasM {
+		w.hasM = true
+		w.precM = precM
+		w.scalings[w.dimensions] = math.Pow10(precM)
+		w.dimensions++
+	}
+
+	if hasZ && hasM {
+		w.ctype = DimXYZM
+	} else if hasZ {
+		w.ctype = DimXYZ
+	} else if hasM {
+		w.ctype = DimXYM
+	} else {
+		w.ctype = DimXY
+	}
+
+	if hasZ || hasM {
+		w.hasExt = true
+	}
```

Comment:
```suggestion
	w.hasExt = hasZ || hasM
```

Reply:
In general that transformed code is not identical (but I think it is valid in this case). I'll make the change.